lib/CL/devices/basic/basic.c

```diff
@@ -719,14 +719,21 @@ pocl_basic_fill_rect (void *data,
                 + buffer_slice_pitch * k, fill_pixel, pixel_size);
 }
 
+typedef unsigned long v2ul __attribute__ ((vector_size (16)));
```

Comment:
Also, do you now wrongly assume the data is aligned to the vector length. The accesses through vector pointers assume natural alignment, at least with Clang. It think the original switch..case optimization might be counter productive too (whoever implemented that). Did you measure the impact on various machines with autovectorizing compilers and CPUs with SIMD instructions?